support-lib/jni/djinni_support.cpp

```diff
@@ -239,8 +239,13 @@ jint JniEnum::ordinal(JNIEnv * env, jobject obj) const {
 
 LocalRef<jobject> JniEnum::create(JNIEnv * env, jint value) const {
     LocalRef<jobject> values(env, env->CallStaticObjectMethod(m_clazz.get(), m_staticmethValues));
+    jniExceptionCheck(env);
     DJINNI_ASSERT(values, env);
-    return LocalRef<jobject>(env, env->GetObjectArrayElement(static_cast<jobjectArray>(values.get()), value));
+    LocalRef<jobject> result(env,
+                             env->GetObjectArrayElement(static_cast<jobjectArray>(values.get()),
+                                                        value));
+    jniExceptionCheck(env);
+    return std::move(result);
```

Comment:
I'm not sure I buy that "declared as const" is a meaningful thing from a compiler's point of view in this case, given there are 2 variable declarations here anyway.  RVO is collapsing them into one variable, explicitly breaking standard guarantees such as different variables having unique addresses.  It could be allowed to break other guarantees too, if that were deemed to be safe and valuable.

To be clear lest we be arguing across purposes.  I totally buy that using RVO in case 1 is illegal, and don't dispute that.  I'm asking why.  So I'm not looking for an answer in the form saying that it's illegal based on more different ways that it is illegal.  I'm looking for a bits-and-bytes justification for why it should be illegal, because if it wasn't compilers would do something unsafe, or be unable to do something desireable.  Because the standard rules are set up because someone thinks they're a good idea, and if an alternative would be clearly better for developers with no negative side-effects the standard probably would've allowed it.  I kinda want a condensed replay of the discussion in the standards committee which lead to this set of rules.  I suspect nobody in this project has that, though. :)